### ISSUE TITLE
Replace github login with email login

### ISSUE THREAD

**Issue by moellep** [radiasoft/sirepo#1173]
Replace the github oauth login with a password-less email login, similar to workflow used by medium.com.

- [ ] rsconf updates to config: oauth => github, feature_api_modules
- [ ] post release update github links for new format (callback_uri update in rsconf, too)

**Comment by bruhwiler** [comment]
Please also require users to enter their name.

**Comment by robnagler** [comment]
We may want to coordinate with the transition to a new cookie. Knowing that the old cookie existed would be a good sentinel for knowing we didn't have their name, etc.

**Comment by moellep** [comment]
We've discussed this a bit and will continue to support the github login in the software, but that login method won't be available on sirepo.com. 

We've also discussed requiring a logged in user before they have access to the simulations. Another solution would be to allow anonymous users to access the example simulations, but get prompted to login if they want to copy/create/import a new simulation. That would allow someone to try the software immediately without needing to wait for the registration email to get through any grey-listing delays.

**Comment by bruhwiler** [comment]
I like the suggested approach.

**Comment by robnagler** [comment]
To be clear, they could still modify the parameters of the example simulation?

**Comment by moellep** [comment]
The email login workflow has been implemented. It follows a similar workflow to the Medium website. A user can sign in with their email address, which sends an email with a special link to their inbox. The authorization link can only be used once and expires after 15 minutes. Clicking the link will log the user into their sirepo workspace. When the user logs in the first time, they will be prompted to enter their full name to complete their registration.


The user selects the email login menu item.
<img width="289" alt="screen shot 2019-01-10 at 3 27 50 pm" src="https://user-images.githubusercontent.com/6516910/51001669-f020d900-14ed-11e9-997c-9995deb0914a.png">

The user is prompted for their email address.
<img width="752" alt="screen shot 2019-01-10 at 3 28 09 pm" src="https://user-images.githubusercontent.com/6516910/51001673-f3b46000-14ed-11e9-9837-26a1f50d98ad.png">

A confirmation page is shown asking the user to check their inbox.
<img width="745" alt="screen shot 2019-01-10 at 3 29 21 pm" src="https://user-images.githubusercontent.com/6516910/51001680-f8791400-14ed-11e9-9ac8-282668892056.png">

The user receives an email with an authorization link.
<img width="672" alt="screen shot 2019-01-10 at 3 29 46 pm" src="https://user-images.githubusercontent.com/6516910/51001686-fc0c9b00-14ed-11e9-9eb0-d707b8e9cbf5.png">

The user follows the link and is logged in. The first visit, the user is prompted for their full name.
<img width="743" alt="screen shot 2019-01-10 at 3 30 01 pm" src="https://user-images.githubusercontent.com/6516910/51001691-ff078b80-14ed-11e9-9e3a-0f2c2406ff26.png">

When a user is logged in, the user icon nav bar is shown in color. Unlike github login, there is no avatar.
<img width="996" alt="screen shot 2019-01-10 at 3 32 05 pm" src="https://user-images.githubusercontent.com/6516910/51001708-0f1f6b00-14ee-11e9-8a7b-216c634d6b97.png">

Selecting the user icon, and the current user email is displayed.
<img width="341" alt="screen shot 2019-01-10 at 3 30 14 pm" src="https://user-images.githubusercontent.com/6516910/51001697-02027c00-14ee-11e9-823c-9b9f4f3f07fa.png">

If an invalid or expired authorization link is followed, an error pages is displayed.
<img width="993" alt="screen shot 2019-01-10 at 3 30 46 pm" src="https://user-images.githubusercontent.com/6516910/51001704-06c73000-14ee-11e9-8e9d-b07954ce5f25.png">
 
When an existing user requests authorization, the email text is slightly different.
<img width="657" alt="screen shot 2019-01-10 at 3 46 54 pm" src="https://user-images.githubusercontent.com/6516910/51001957-024f4700-14ef-11e9-945e-626f068bf602.png">

**Comment by robnagler** [comment]
Master is merged in the email-auth branch. However, I was going through the code, and I just found it very difficult to understand, because of the way anonymous logins are handled. When there are many `if` statements for a special case, the code is hard to follow. Since this is about authentication, it needs to be simpler.

What I believe we want is the following use cases:

* Email auth without the possibility of anonymous use
* GitHub auth with the possibility of anonymous use
* Anonymous use only

In thinking through the code, I think it's a mistake to not at least ask for a name, even in the anonymous case. If someone doesn't want to give a name, they can enter "x" or whatever. I'm going to create a "guest" auth that will act the same as anonymous except that it will require a name to display and, more importantly, to help people with who have 3rd party installations. 

GitHub will also require a name.  oauth.py has if statements that are dubious, too, because there's only one oauth method. Therefore, oauth.py will become github_auth.py. If we implement another oauth method, we can share the code.

Right now, the email-auth branch supports the concept of a login_module to handle oauth or email_auth. Inside the cookie, there's an auth_method value which maps to either of these modules. There's code to allow anonymous logins with oauth. What we really want is a list of auth methods. For development, the method list will be "guest" (default). For GitHub, it will be either `github` or `guest`. For email auth (sirepo.com), it'll be `email`. 

With email auth we have a special case to convert from GitHub to Email. That looks at the auth method and forces it. However, this is yet another special case. The general case is that if you are logged out, you'll still have an auth method in the cookie, which tells the app which method you were logged in as. You'll be forced to login via GitHub when you visit the site. If the only auth method that's available is email, then you'll be forced to login again to register your email and migrate the user from GitHub. (This code is written as a special case.) This will allow sites to migrate to new auth methods seamlessly.

When you logout as guest, your user will be cleared completely. That's what we want for development, because we have to create new users, and that's really what happens when you logout. 

The GUI currently asks you how to login after you login. This needs to change to be a login choice when you come into the app. If there is only one choice, it will go directly to that login modal. If there's a choice, you'll get "login with GitHub" or "login as guest" choices, which is where you'll end up when you logout so really nothing will change except you'll get hit with the question first time you visit the site. Once logged in, you stay stay logged in.
